### ISSUE TITLE
do not ignore files started with `.` by default

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#10341]
**The version of ESLint you are using.**
master
**The problem you want to solve.**
eslint ingore `.eslintrc.js` by default, so I have to add a file `.eslintignore`, then add `!.eslintrc.js` in. 

**Your take on the correct solution to problem.**
do not ignore file/dir beginning with a dot by default, except for `.` and `..`(not sure `.git` can be exception)

In some cases ignoring is a good idea, while ingoring all `.*` files seems an overkill.

**Comment by Aladdin-ADD** [comment]
this behavior seems not documented. am I missing something?

https://eslint.org/docs/user-guide/configuring#ignoring-files-and-directories

**Comment by not-an-aardvark** [comment]
Related: https://github.com/eslint/eslint/issues/10089

**Comment by platinumazure** [comment]
This will need to be discussed by the TSC. But before we can discuss, I think we need a concrete proposal. (i.e., exactly what files should we ignore by default, or not?)

**Comment by kaicataldo** [comment]
I've been bitten by this in the past too, and I agree that it's unintuitive. I would support this change! Funnily enough, I just had to fix this in [a project](https://github.com/kaicataldo/css-modules-ts-definitions-loader/blob/master/.eslintignore) today.

**Comment by Aladdin-ADD** [comment]
@platinumazure , to be clearer, I am proposing: do not ignore file/dir beginning with a dot by default, except for `.` and `..`
(not sure `.git`can be exception)

**Comment by platinumazure** [comment]
I think this needs to be fleshed out a little more. There's not much for the TSC to review right now. (Plus, we usually like a [summary and question](https://eslint.org/docs/maintainer-guide/issues#when-to-send-to-tsc) to be provided so we have a good starting point.)

@Aladdin-ADD @kaicataldo Could one of you add some context/flesh out a proposal? Thanks!

**Comment by Aladdin-ADD** [comment]
**TSC Summary**: eslint ingore `.*` by default -- it can cause confusing to users. so I am proposing: do not ignore file/dir beginning with a dot by default, except `.` and `..`

**TSC Question**: 
+ Should we accept this change?
+ Should `.git` be ignored by default?

**Comment by kaicataldo** [comment]
@Aladdin-ADD 

Just so I understand, can you explain the `except . and ..` part for me? Won't the shell always expand this?

**Comment by Aladdin-ADD** [comment]
`.` and `..` are special cases in `.*`, I just hope they can be treated as is. please help to better the proposal.  thanks!😄 @kaicataldo

**Comment by mysticatea** [comment]
I think, plus `.hg` and `.svn` as well.

**Comment by platinumazure** [comment]
I'm worried about the potential number of exceptions here. I personally lean towards ignoring those by default (after all, they're hidden from the user and usually for a good reason), but making sure we are equipping the user to handle these cases.

My position:
1. Users will be inconvenienced more by auto linting of dot folders/files and node_modules, so we should ignore by default to simplify things for new users. (We should very clearly document what we ignore by default and how to unignore it.)
1. Users should be able to use ignore patterns like `!.*`, `!.*/**`, etc. to unignore dot files or folders if needed. Is ESLint allowing this? Or is ESLint ignoring those unignores? If it is not handling the unignores, I think we need to change that.
1. Users should be able to see warnings when files or folders are ignored. Are we showing warnings consistently for files ignored by .eslintignore, vs dot folders and files? If not, we should be.

**Comment by kaicataldo** [comment]
> Users should be able to use ignore patterns like !.*, !.*/**, etc. to unignore dot files or folders if needed. Is ESLint allowing this? Or is ESLint ignoring those unignores? If it is not handling the unignores, I think we need to change that.

This is the current behavior and what I've been doing in my own codebases to ensure I'm writing valid JS in config files

**Comment by mysticatea** [comment]
In my case is [VuePress](https://vuepress.vuejs.org/) (I have written some documentation of my OSS with VuePress). I write many `*.js` and `*.vue` files in `.vuepress` directory, but ESLint ignores those by default. Plus, I want to check `.eslintrc.js` as well.

I feel odd if I have to use `.eslintignore` to not ignore files.
I feel good if I have to use `.eslintignore` to ignore files.

**Comment by ilyavolodin** [comment]
It's a very common behavior to ignore files starting with dot. Even bash ignores them by default when you listing a directory. I'm not sure it's unexpected and we do provide a work-around to include them (although, I do agree with @mysticatea it's a bit confusing to use ignore file to unignore patterns).

**Comment by not-an-aardvark** [comment]
In [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/430455a12366c4614f322734b38521ef12870251/notes/2018/2018-08-02.md#do-not-ignore-files-started-with--by-default), the TSC decided not to make a change for now, but to improve the documentation about existing options for how to unignore dotfiles (e.g. with the `--ignore-pattern` option). We plan to reconsider the issue around the next major release, since we wouldn't be able to change the default until then anyway.

**Comment by rightaway** [comment]
@not-an-aardvark How could we use `--ignore-pattern` to unignore dotfiles?

How could I add feedback to the TSC about this issue? There's a couple of relevant use cases I'd like to bring up related to the specific points in the note from that meeting.

**Comment by not-an-aardvark** [comment]
@rightaway You can use `--ignore-pattern='!.*'` to unignore dotfiles. 

If you have feedback, feel free to provide it in a comment on this issue.

**Comment by nzakas** [comment]
As this is a breaking change, it falls under our new RFC process. To move this forward, someone should put together an RFC. Details can be found here:
https://github.com/eslint/rfcs

Thanks!

**Comment by aladdin-add** [comment]
I strongly suggest to accept it.

`/node_modules/*` and `/bower_components/*` were ignored by default, since nearly all devs do not want to lint them. But `".*"` is not the case -- it was widely used as config file in the community -- I image a large number of users didn't even notice some files in their repo not linting.

at the very least, we need to document it.

**Comment by not-an-aardvark** [comment]
I agree that at the very least we should document it.

Personally, I think linting all dotfiles would lead to confusion because (e.g.) files in `.git` could be linted. I think it could be reasonable to make a special exception and not ignore `.eslintrc.js` files.

(In [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/3e40976a2ebb9481405d4c3a358bc3191ecd234a/notes/2019/2019-02-28.md#do-not-ignore-files-started-with--by-default), the TSC decided to table this issue for v6.0.0.)

**Comment by jaydenseric** [comment]
It's better to blacklist, rather than whitelist, dotfiles from being linted.

People can tell right away when something undesired is being linted; and it intuitive and easy to add it to their ESLint ignore config. The other way around is a hassle.

For example, if you want to lint dotfiles such as `.babelrc` you not only have to un-ignore the dotfiles, but re-ignore `node_modules` afterwards, otherwise any dotfiles in `node_modules` would get linted too 😓

https://github.com/jaydenseric/graphql-react/blob/v8.1.1/package.json#L108

```json
{
  "eslintIgnore": ["!.*", "node_modules"]
}
```

It's weird that you use an ignore list to un-ignore things.

Heaps of tools use [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) for config files, so there are hundreds of variations of `.examplerc` and `.examplerc.js` out in the wild.

**Comment by aladdin-add** [comment]
closing, as it looks like this issue didn't get consensus from the team.

**Comment by brandonchinn178** [comment]
Hello! I'm hitting this now in my project:
```
$ yarn eslint client/.storybook/config.js
.../client/.storybook/config.js
  0:0  warning  File ignored because of a matching ignore pattern. Use "--no-ignore" to override
```

I see this comment: https://github.com/eslint/eslint/issues/4828#issuecomment-194535273

> Thanks. So this looks like a bug, we should be linting anything passed on the command line. We should fix that.

But it seems like that's not the case? https://github.com/eslint/eslint/blob/22046939a7f250e6cce40a2acdfd3581a2ce29f5/tests/lib/cli-engine.js#L355-L375

The documentation also doesn't seem helpful. Searching for "dotfiles" doesn't help me, and [the section on `.eslintignore`](https://eslint.org/docs/user-guide/configuring#eslintignore) only mentions

> Lines preceded by ! are negated patterns that re-include a pattern that was ignored by an earlier pattern.

(I'm not ignoring anything)

It's also weird that the message says `File ignored because of a matching ignore pattern` because I don't have a `.eslintignore` file in my project.

The weird part is that `yarn eslint client/.storybook/` works.

**Comment by kaicataldo** [comment]
Feel free to make a new issue with the problem you’re encountering!
